File: solutions/two-pointers/026-remove-duplicates-from-sorted-array.py
"""
# Difficulty: Easy

# 026. Remove Duplicates From Sorted Array

Given an integer array nums sorted in non-decreasing order, remove the duplicates in-place such that each unique element appears only once. The relative order of the elements should be kept the same.

Since it is impossible to change the length of the array in some languages, you must instead have the result be placed in the first part of the array nums. More formally, if there are k elements after removing the duplicates, then the first k elements of nums should hold the final result.

Return k after placing the final result in the first k slots of nums.

**Example:**

<dl class="example-details">
<dt>Input:</dt>
<dd>[1,1,2]</dd>
<dt>Output:</dt>
<dd>1</dd>
<dt>Explanation:</dt>
<dd>Remove duplicates from sorted array [1,1,2] gives length 2</dd>
</dl>

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### INTUITION:
This is a classic two-pointers problem. Since the array is sorted, duplicates are adjacent. We use one pointer to iterate through the array and another to track the position where the next unique element should be placed.

### APPROACH:
1. **Two pointers**: Use `i` to iterate and `j` to track unique position
2. **Skip duplicates**: Only advance `j` when we find a new unique element
3. **In-place modification**: Copy unique elements to positions 0, 1, 2, etc.
4. **Return count**: Return the number of unique elements

### WHY THIS WORKS:
- Sorted array means duplicates are adjacent
- Two pointers allow in-place removal without extra space
- `j` tracks the "write" position for next unique element
- `i` scans through all elements

### EXAMPLE WALKTHROUGH:
```
Input: [1,1,2]
i=0, j=0: nums[0]=1 (first element, place at j=0)
i=1, j=1: nums[1]=1 == nums[0], skip
i=2, j=1: nums[2]=2 != nums[0], place at j=1
Result: [1,2,_], return k=2
```

### TIME COMPLEXITY:
O(n)
Single pass through the array

### SPACE COMPLEXITY:
O(1)
Only using constant extra space

### EDGE CASES:
- **Empty array**: Return 0 (no elements)
- **Single element**: Return 1 (already unique)
- **All elements same**: Return 1 (only one unique value)
- **No duplicates**: Return n (all unique already)
- **Consecutive duplicates**: Two-pointer removes them in-place

</details>
"""


class Solution:
    def removeDuplicates(self, nums: list[int]) -> int:
        """
        Remove duplicates from sorted array in-place using two pointers.

        Args:
            nums: Sorted array in non-decreasing order

        Returns:
            Number of unique elements (length of modified array)

        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        if not nums:
            return 0

        # j tracks the position for next unique element
        j = 1

        for i in range(1, len(nums)):
            # If current element is different from previous unique element
            if nums[i] != nums[j - 1]:
                nums[j] = nums[i]
                j += 1

        return j

    def removeDuplicatesAlternative(self, nums: list[int]) -> int:
        """
        Alternative implementation with explicit unique tracking.

        Args:
            nums: Sorted array

        Returns:
            Number of unique elements
        """
        if not nums:
            return 0

        unique_count = 1  # First element is always unique

        for i in range(1, len(nums)):
            if nums[i] != nums[unique_count - 1]:
                nums[unique_count] = nums[i]
                unique_count += 1

        return unique_count


def test_solution() -> None:
    """Test cases for 026. Remove Duplicates From Sorted Array."""
    solution = Solution()

    # Test case 1: Example case
    nums1 = [1, 1, 2]
    result1 = solution.removeDuplicates(nums1)
    expected1 = 2
    assert result1 == expected1 and nums1[:result1] == [1, 2], f"Expected {expected1}, got {result1}"

    # Test case 2: More duplicates
    nums2 = [0, 0, 1, 1, 1, 2, 2, 3, 3, 4]
    result2 = solution.removeDuplicatesAlternative(nums2)
    expected2 = 5
    assert result2 == expected2 and nums2[:result2] == [0, 1, 2, 3, 4], f"Expected {expected2}, got {result2}"

    # Test case 3: No duplicates
    nums3 = [1, 2, 3]
    result3 = solution.removeDuplicates(nums3)
    expected3 = 3
    assert result3 == expected3, f"Expected {expected3}, got {result3}"

    # Test case 4: All same
    nums4 = [1, 1, 1]
    result4 = solution.removeDuplicates(nums4)
    expected4 = 1
    assert result4 == expected4, f"Expected {expected4}, got {result4}"

    print("All test cases passed!")


if __name__ == "__main__":
    test_solution()

    # Example usage
    solution = Solution()
    nums = [1, 1, 2, 2, 3]
    k = solution.removeDuplicates(nums)
    print(f"Unique elements: {k}, Array: {nums[:k]}")

File: solutions/two-pointers/alternatives/026-remove-duplicates-from-sorted-array.js
/**
 * Difficulty: Medium
 *
 * [Problem description goes here]
 *
 * **Example:**
 *
 * <dl class="example-details">
 * <dt>Input:</dt>
 * <dd>[input description]</dd>
 * <dt>Output:</dt>
 * <dd>[output description]</dd>
 * <dt>Explanation:</dt>
 * <dd>[explanation]</dd>
 * </dl>
 *
 * <details>
 * <summary><b>🔍 SOLUTION EXPLANATION</b></summary>
 *
 * ### INTUITION:
 * [High-level insight or key observation]
 *
 * ### APPROACH:
 * [Detailed explanation of the solution approach]
 *
 * ### WHY THIS WORKS:
 * - [Explanation of correctness]
 *
 * ### EXAMPLE WALKTHROUGH:
 * Input:
 * ```
 * [example input]
 * ```
 *
 * **Step 1:** [description]
 *
 * **Step 2:** [description]
 *
 * ### TIME COMPLEXITY:
 * **O(?)** - [explanation]
 *
 * ### SPACE COMPLEXITY:
 * **O(?)** - [explanation]
 *
 * ### EDGE CASES:
 * - **[Edge case 1]:** [how it's handled]
 * - **[Edge case 2]:** [how it's handled]
 *
 * </details>
 */

/**
 * Main solution for Problem 026: Remove Duplicates From Sorted Array
 *
 * @param {number[]} nums - Sorted array of integers
 * @return {number} - Number of unique elements
 *
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */
function solve(nums) {
    // Edge case: empty array
    if (!nums || nums.length === 0) {
        return 0;
    }

    // Slow pointer tracks position for next unique element
    let slow = 0;

    // Fast pointer scans through array
    for (let fast = 1; fast < nums.length; fast++) {
        // Found a new unique element
        if (nums[fast] !== nums[slow]) {
            slow++;
            nums[slow] = nums[fast];
        }
    }

    // Return count of unique elements (slow is 0-indexed, so add 1)
    return slow + 1;
}

/**
 * Test cases for Problem 026: Remove Duplicates From Sorted Array
 */
function testSolution() {
    console.log('Testing 026. Remove Duplicates From Sorted Array');

    // Test case 1: Example from problem
    const nums1 = [1, 1, 2];
    const result1 = solve(nums1);
    const expected1 = 2;
    const expectedArray1 = [1, 2];
    console.assert(result1 === expected1, `Test 1 failed: expected ${expected1}, got ${result1}`);
    console.assert(nums1.slice(0, result1).toString() === expectedArray1.toString(),
        `Test 1 failed: array should be ${expectedArray1}, got ${nums1.slice(0, result1)}`);
    console.log(`Test 1 passed: removeDuplicates([1,1,2]) = ${result1}, nums = [${nums1.slice(0, result1)}]`);

    // Test case 2: Multiple duplicates
    const nums2 = [0, 0, 1, 1, 1, 2, 2, 3, 3, 4];
    const result2 = solve(nums2);
    const expected2 = 5;
    const expectedArray2 = [0, 1, 2, 3, 4];
    console.assert(result2 === expected2, `Test 2 failed: expected ${expected2}, got ${result2}`);
    console.assert(nums2.slice(0, result2).toString() === expectedArray2.toString(),
        `Test 2 failed: array should be ${expectedArray2}, got ${nums2.slice(0, result2)}`);
    console.log(`Test 2 passed: removeDuplicates([0,0,1,1,1,2,2,3,3,4]) = ${result2}, nums = [${nums2.slice(0, result2)}]`);

    // Test case 3: All same elements
    const nums3 = [1, 1, 1, 1];
    const result3 = solve(nums3);
    const expected3 = 1;
    console.assert(result3 === expected3, `Test 3 failed: expected ${expected3}, got ${result3}`);
    console.log(`Test 3 passed: removeDuplicates([1,1,1,1]) = ${result3}`);

    // Test case 4: No duplicates
    const nums4 = [1, 2, 3, 4, 5];
    const result4 = solve(nums4);
    const expected4 = 5;
    console.assert(result4 === expected4, `Test 4 failed: expected ${expected4}, got ${result4}`);
    console.log(`Test 4 passed: removeDuplicates([1,2,3,4,5]) = ${result4}`);

    // Test case 5: Single element
    const nums5 = [1];
    const result5 = solve(nums5);
    const expected5 = 1;
    console.assert(result5 === expected5, `Test 5 failed: expected ${expected5}, got ${result5}`);
    console.log(`Test 5 passed: removeDuplicates([1]) = ${result5}`);

    console.log('All test cases passed for 026. Remove Duplicates From Sorted Array!');
}

/**
 * Example usage and demonstration
 */
function demonstrateSolution() {
    console.log('\n=== Problem 026. Remove Duplicates From Sorted Array ===');
    console.log('Category: Two Pointers');
    console.log('Difficulty: Easy');
    console.log('');

    // Example demonstration would go here
    testSolution();
}

// Run tests if this file is executed directly
if (require.main === module) {
    demonstrateSolution();
}

// Export for use in other modules
module.exports = {
    solve,
    testSolution,
    demonstrateSolution
};

/**
 * Additional Notes:
 * - This solution focuses on two pointers concepts
 * - The slow-fast pointer pattern is a common technique for in-place array modifications
 * - Edge cases are crucial for robust solutions
 * - The approach can be adapted for similar problems in this category
 */
